File: Algorithms/p341_Flatten_Nested_List_Iterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger(object):
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """

class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.nested_list = nestedList
        self.index = -1
        self.my_stack = []
        self.nextHelper()

    def nextHelper(self):
        while True:
            while self.my_stack:
                self.my_stack[-1][1] += 1
                if(self.my_stack[-1][1]!=len(self.my_stack[-1][0].getList())):
                    break
                self.my_stack.pop()
            if not self.my_stack:
                self.index += 1
                if self.index==len(self.nested_list):
                    break
                self.my_stack.append([self.nested_list[self.index],0])
            while not self.my_stack[-1][0].isInteger() and self.my_stack[-1][0].getList():
                self.my_stack.append([self.my_stack[-1][0].getList()[self.my_stack[-1][1]],0])
            if self.my_stack[-1][0].isInteger():
                break
            else:
                self.my_stack.pop()

    def next(self):
        """
        :rtype: int
        """
        res = self.my_stack[-1][0].getInteger()
        self.my_stack.pop()
        self.nextHelper()
        return res

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.index!=len(self.nested_list)

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: Algorithms/p341_Flatten_Nested_List_Iterator/p341_Flatten_Nested_List_Iterator.cpp
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */
class NestedIterator {
    vector<NestedInteger>* nested_list;
    int index;
    stack<pair<NestedInteger*,int>,vector<pair<NestedInteger*,int>>> my_stack;

    void nextHelper(){
        while(1){
            while(!my_stack.empty()){
                if(++my_stack.top().second!=my_stack.top().first->getList().size()) break;
                my_stack.pop();
            }
            if(my_stack.empty()){
                if(++index==nested_list->size()) break;
                my_stack.push(make_pair(&(*nested_list)[index],0));
            }
            while(!my_stack.top().first->isInteger()&&!my_stack.top().first->getList().empty()){
                my_stack.push(make_pair(&my_stack.top().first->getList()[my_stack.top().second],0));
            }
            if(my_stack.top().first->isInteger()) break;
            else my_stack.pop();
        }
    }

public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        nested_list = &nestedList;
        index = -1;
        nextHelper();
    }

    int next() {
        int res = my_stack.top().first->getInteger();
        my_stack.pop();
        nextHelper();
        return res;
    }

    bool hasNext() {
        return index!=nested_list->size();
    }
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */
